File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:

	int lengthOfLongestSubstring(std::string s) {
		
		std::queue<char> window = {};
		int max_len = 0;
		std::unordered_set<char> letter_set;

		for (int i = 0; i < s.size(); i++) {

			while (letter_set.find(s[i]) != letter_set.end()) {
				letter_set.erase(window.front());
                window.pop();
				
			}

			window.push(s[i]);
			letter_set.insert(s[i]);
			max_len = std::max(max_len, (int)window.size());


		}

		return max_len;


	}


};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        
        n = len(s)
        word_set = set()
        window = []
        
        max_len = 0
        
        for i in range(0,n):
            while s[i] in word_set:
                word_set.remove(window[0])
                window.pop(0)
            window.append(s[i])
            word_set.add(s[i])
            max_len = max(max_len, len(window))
        
        
        return max_len
                
            
        
        
        
        